### PR TITLE
Add Option to use NotFoundHttpException for 404 HTTP errors

### PR DESCRIPTION
**Pull request by blankse**
Issue: https://github.com/ezsystems/LegacyBridge/issues/87

- [X] Unit test
- [x] Manual test

### CODE REVIEW

**Comment by emodric** [comment]
Looks good to me :+1:

**Comment by blankse** [comment]
@kemoc @lolautruche Can you review this?

**Comment by kemoc** [comment]
@blankse seems legit

**Comment by lolautruche** [comment]
This should be documented somehow.
I'm not a fan of having a setting for that but I don't see another way to keep BC without it.

Any idea @bdunogier @andrerom ?

**Comment by andrerom** [comment]
@lolautruche well, I guess the new behavior is what is should be, it's after all non legacy mode. What about introducing that as the default and only behavior as of 1.3.0?

**Comment by lolautruche** [comment]
> What about introducing that as the default and only behavior as of 1.3.0

I don't agree, since this would be a BC break. This should go to 2.0.
(And the fix should be backported to 5.4 IMHO)

**Comment by emodric** [comment]
I'm with @lolautruche on this one.

**Comment by emodric** [comment]
What we can do is add an `E_USER_DEPRECATED` in case legacy kernel handles the exception.

File: bundle/LegacyResponse/LegacyResponseManager.php [hunk 1]
```diff
@@ -87,11 +97,27 @@ public function generateResponseFromModuleResult(ezpKernelResult $result)
...
+                    $errorMessage = isset($moduleResult['errorMessage']) ? $moduleResult['errorMessage'] : 'Access denied';
+                    throw new AccessDeniedException($errorMessage);
+                }
+                // If having an "Not found" error code in non-legacy mode and conversation is true,
+                // we send an NotFoundHttpException to be able to trigger error page in Symfony stack.
+                if ($moduleResult['errorCode'] == 404) {
+                    if ($this->notFoundHttpConversion) {
+                        $errorMessage = isset($moduleResult['errorMessage']) ? $moduleResult['errorMessage'] : 'Not found';
+                        throw new NotFoundHttpException($errorMessage);
+                    } else {
```
Comment (c-revillo) [inline bundle/LegacyResponse/LegacyResponseManager.php]: is the `else` needed? i mean, you are throwing the exception so maybe else can be avoided...

Comment (blankse) [inline bundle/LegacyResponse/LegacyResponseManager.php]: You are right. I removed the else ;)

Comment (c-revillo) [inline bundle/LegacyResponse/LegacyResponseManager.php]: :+1:

**Comment by blankse** [comment]
@emodric I added an E_USER_DEPRECATED if legacy handles the 404 error.

File: bundle/LegacyResponse/LegacyResponseManager.php [hunk 2]
```diff
@@ -106,9 +106,16 @@ public function generateResponseFromModuleResult(ezpKernelResult $result)
...
-                    $errorMessage = isset($moduleResult['errorMessage']) ? $moduleResult['errorMessage'] : 'Not found';
-                    throw new NotFoundHttpException($errorMessage);
+                if ($moduleResult['errorCode'] == 404) {
+                    if ($this->notFoundHttpConversion) {
+                        $errorMessage = isset($moduleResult['errorMessage']) ? $moduleResult['errorMessage'] : 'Not found';
+                        throw new NotFoundHttpException($errorMessage);
+                    }
+                    @trigger_error(
+                        "Legacy 404 error handling is deprecated, and will be removed in legacy-bridge 2.0.\n" .
+                        'Use the not_found_http_conversion setting to use the new behavior yet.',
```
Comment (emodric) [inline bundle/LegacyResponse/LegacyResponseManager.php]: `to use the new behavior and disable this notice`

**Comment by emodric** [comment]
@blankse +1 With an inline nitpick fixed

File: bundle/LegacyResponse/LegacyResponseManager.php (hunk 2 above)
Comment (blankse) [inline bundle/LegacyResponse/LegacyResponseManager.php]: I changed it

**Pull request merged**